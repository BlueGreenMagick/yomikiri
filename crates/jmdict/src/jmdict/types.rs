use serde::{Deserialize, Serialize};

use crate::utils::jm_entity_enum;

#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMDict {
    pub entries: Vec<JMEntry>,
    pub creation_date: Option<String>,
}

#[derive(Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMEntry {
    pub id: u32,
    /// 0+ k_ele
    pub kanjis: Vec<JMKanji>,
    /// 1+ r_ele
    pub readings: Vec<JMReading>,
    /// 1+ sense
    pub senses: Vec<JMSense>,
}

impl JMEntry {
    pub fn priority(&self) -> u16 {
        let priorities = &self.readings.first().unwrap().priority;
        let mut priority: u16 = 0;

        for p in priorities {
            // common ~20k entries
            if ["news1", "ichi1", "spec1", "gai1"].contains(&p.as_str()) {
                if priority < 100 {
                    priority += 100
                } else {
                    priority += 25
                }
            // common ~30k entries
            } else if ["news2", "ichi2", "spec2", "gai2"].contains(&p.as_str()) {
                priority += 5
            // 01 ~ 48, each with ~500 entries
            } else if let Some(stripped) = p.strip_prefix("nf") {
                let freq = stripped
                    .parse::<u16>()
                    .expect("could not parse XX as number where priority nfXX");
                priority += 50 - freq;
            }
        }
        priority
    }
}

#[derive(Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMKanji {
    pub kanji: String,
    /// 0+ ke_inf
    pub info: Vec<JMKanjiInfo>,
    /// 0+ ke_pri
    pub priority: Vec<String>,
}

impl JMKanji {
    pub fn is_uncommon(&self) -> bool {
        self.info.iter().any(|s| {
            *s == JMKanjiInfo::OutdatedKanji
                || *s == JMKanjiInfo::RareKanjiForm
                || *s == JMKanjiInfo::SearchOnlyKanji
        })
    }
}

#[derive(Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMReading {
    pub reading: String,
    pub nokanji: bool,
    /// 0+ re_restr
    pub to_form: Vec<String>,
    /// 0+ re_inf
    pub info: Vec<JMReadingInfo>,
    /// 0+ re_pri
    pub priority: Vec<String>,
}

impl JMReading {
    pub fn is_uncommon(&self) -> bool {
        self.info
            .iter()
            .any(|s| *s == JMReadingInfo::Outdated || *s == JMReadingInfo::SearchOnly)
    }
}

#[derive(Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMSense {
    /// 0+ 'stagk'
    pub to_form: Vec<String>,
    /// 0+ 'stagr'
    pub to_reading: Vec<String>,
    /// 'pos'
    pub pos: Vec<JMPartOfSpeech>,
    /// 'xref'
    // pub reference: Vec<String>,
    /// 'ant'
    // pub antonym: Vec<String>,
    /// 'field'
    // pub field: Vec<String>,
    /// 'misc'
    pub misc: Vec<JMSenseMisc>,
    /// 's_inf'
    pub info: Vec<String>,
    /// 'dial'
    pub dialects: Vec<JMDialect>,
    /// 'gloss'
    pub meanings: Vec<String>,
    // 'example'
    // example: Vec<Example>,
}

jm_entity_enum!(
    JMDialect;
    b"hob" => HokkaidoBen,
    b"bra" => Brazilian,
    b"ksb" => KansaiBen,
    b"ktb" => KantouBen,
    b"kyb" => KyotoBen,
    b"kyu" => KyuushuuBen,
    b"nab" => NaganoBen,
    b"osb" => OsakaBen,
    b"rkb" => RyuukyuuBen,
    b"thb" => TouhokuBen,
    b"tsb" => TosaBen,
    b"tsug" => TsugaruBen,
);

jm_entity_enum!(
    JMKanjiInfo;
    b"ateji" => AtejiReading,
    b"ik" => IrregularKana,
    b"iK" => IrregularKanji,
    b"io" => IrregularOkurigana,
    b"oK" => OutdatedKanji,
    b"rK" => RareKanjiForm,
    b"sK" => SearchOnlyKanji,
);

jm_entity_enum!(
    JMSenseMisc;
    b"abbr" => Abbreviation,
    b"arch" => Archaic,
    b"char" => Character,
    b"chn" => Children,
    b"col" => Colloquial,
    b"company" => Company,
    b"creat" => Creature,
    b"dated" => DatedTerm,
    b"dei" => Deity,
    b"derog" => Derogatory,
    b"doc" => Document,
    b"euph" => Euphemistic,
    b"ev" => Event,
    b"fam" => Familiar,
    b"fem" => Female,
    b"fict" => Fiction,
    b"form" => FormalTerm,
    b"given" => Forename,
    b"group" => Group,
    b"hist" => HistoricalTerm,
    b"hon" => Honorific,
    b"hum" => Humble,
    b"id" => Idiomatic,
    b"joc" => Jocular,
    b"leg" => Legend,
    b"m-sl" => MangaSlang,
    b"male" => Male,
    b"myth" => Mythology,
    b"net-sl" => InternetSlang,
    b"obj" => Object,
    b"obs" => Obsolete,
    b"on-mim" => Onomatopoeic,
    b"organization" => Organization,
    b"oth" => Other,
    b"person" => Person,
    b"place" => Place,
    b"poet" => Poetical,
    b"pol" => Polite,
    b"product" => Product,
    b"proverb" => Proverb,
    b"quote" => Quotation,
    b"rare" => Rare,
    b"relig" => Religion,
    b"sens" => Sensitive,
    b"serv" => Service,
    b"ship" => Ship,
    b"sl" => Slang,
    b"station" => Railway,
    b"surname" => Surname,
    b"uk" => UsuallyKanaAlone,
    b"unclass" => Unclassified,
    b"vulg" => Vulgar,
    b"work" => Artwork,
    b"X" => RudeTerm,
    b"yoji" => Yojijukugo,
);

jm_entity_enum!(
    JMPartOfSpeech;
    #[doc="noun or verb acting prenominally"]
    b"adj-f" => PrenominalNounOrVerb,
    #[doc="adjective (keiyoushi)"]
    b"adj-i" => Adjective,
    #[doc="adjective (keiyoushi) - yoi/ii class"]
    b"adj-ix" => AdjectiveYoiOrIi,
    #[doc="'kari' adjective (archaic)"]
    b"adj-kari" => AdjectiveKari,
    #[doc="'ku' adjective (archaic)"]
    b"adj-ku" => AdjectiveKu,
    #[doc="adjectival nouns or quasi-adjectives (keiyodoshi)"]
    b"adj-na" => NaAdjectivalNoun,
    #[doc="archaic/formal form of na-adjective"]
    b"adj-nari" => NaAdjectiveNari,
    #[doc="nouns which may take the genitive case particle 'no'"]
    b"adj-no" => NounNo,
    #[doc="pre-noun adjectival (rentaishi)"]
    b"adj-pn" => PrenounAdjectival,
    #[doc="'shiku' adjective (archaic)"]
    b"adj-shiku" => AdjectiveShiku,
    #[doc="'taru' adjective"]
    b"adj-t" => AdjectiveTaru,
    #[doc="adverb (fukushi)"]
    b"adv" => Adverb,
    #[doc="adverb taking the 'to' particle"]
    b"adv-to" => AdverbTo,
    #[doc="auxiliary"]
    b"aux" => Auxiliary,
    #[doc="auxiliary adjective"]
    b"aux-adj" => AuxiliaryAdjective,
    #[doc="auxiliary verb"]
    b"aux-v" => AuxiliaryVerb,
    #[doc="conjunction"]
    b"conj" => Conjunction,
    #[doc="copula"]
    b"cop" => Copula,
    #[doc="counter"]
    b"ctr" => Counter,
    #[doc="expressions (phrases, clauses, etc.)"]
    b"exp" => Expression,
    #[doc="interjection (kandoushi)"]
    b"int" => Interjection,
    #[doc="noun (common) (futsuumeishi)"]
    b"n" => Noun,
    #[doc="adverbial noun (fukushitekimeishi)"]
    b"n-adv" => AdverbialNoun,
    #[doc="proper noun"]
    b"n-pr" => ProperNoun,
    #[doc="noun, used as a prefix"]
    b"n-pref" => PrefixNoun,
    #[doc="noun, used as a suffix"]
    b"n-suf" => SuffixNoun,
    #[doc="noun (temporal) (jisoumeishi)"]
    b"n-t" => TemporalNoun,
    #[doc="numeric"]
    b"num" => Numeric,
    #[doc="pronoun"]
    b"pn" => Pronoun,
    #[doc="prefix"]
    b"pref" => Prefix,
    #[doc="particle"]
    b"prt" => Particle,
    #[doc="suffix"]
    b"suf" => Suffix,
    #[doc="unclassified"]
    b"unc" => Unclassified,
    #[doc="verb unspecified"]
    b"v-unspec" => UnspecifiedVerb,
    #[doc="Ichidan verb"]
    b"v1" => VerbIchidan,
    #[doc="Ichidan verb - kureru special class"]
    b"v1-s" => VerbIchidanKureru,
    #[doc="Nidan verb with 'u' ending (archaic)"]
    b"v2a-s" => VerbNidanU,
    #[doc="Nidan verb (upper class) with 'bu' ending (archaic)"]
    b"v2b-k" => VerbNidanBK,
    #[doc="Nidan verb (lower class) with 'bu' ending (archaic)"]
    b"v2b-s" => VerbNidanBS,
    #[doc="Nidan verb (upper class) with 'dzu' ending (archaic)"]
    b"v2d-k" => VerbNidanDK,
    #[doc="Nidan verb (lower class) with 'dzu' ending (archaic)"]
    b"v2d-s" => VerbNidanDS,
    #[doc="Nidan verb (upper class) with 'gu' ending (archaic)"]
    b"v2g-k" => VerbNidanGK,
    #[doc="Nidan verb (lower class) with 'gu' ending (archaic)"]
    b"v2g-s" => VerbNidanGS,
    #[doc="Nidan verb (upper class) with 'hu/fu' ending (archaic)"]
    b"v2h-k" => VerbNidanHK,
    #[doc="Nidan verb (lower class) with 'hu/fu' ending (archaic)"]
    b"v2h-s" => VerbNidanHS,
    #[doc="Nidan verb (upper class) with 'ku' ending (archaic)"]
    b"v2k-k" => VerbNidanKK,
    #[doc="Nidan verb (lower class) with 'ku' ending (archaic)"]
    b"v2k-s" => VerbNidanKS,
    #[doc="Nidan verb (upper class) with 'mu' ending (archaic)"]
    b"v2m-k" => VerbNidanMK,
    #[doc="Nidan verb (lower class) with 'mu' ending (archaic)"]
    b"v2m-s" => VerbNidanMS,
    #[doc="Nidan verb (lower class) with 'nu' ending (archaic)"]
    b"v2n-s" => VerbNidanNS,
    #[doc="Nidan verb (upper class) with 'ru' ending (archaic)"]
    b"v2r-k" => VerbNidanRK,
    #[doc="Nidan verb (lower class) with 'ru' ending (archaic)"]
    b"v2r-s" => VerbNidanRS,
    #[doc="Nidan verb (lower class) with 'su' ending (archaic)"]
    b"v2s-s" => VerbNidanSS,
    #[doc="Nidan verb (upper class) with 'tsu' ending (archaic)"]
    b"v2t-k" => VerbNidanTK,
    #[doc="Nidan verb (lower class) with 'tsu' ending (archaic)"]
    b"v2t-s" => VerbNidanTS,
    #[doc="Nidan verb (lower class) with 'u' ending and 'we' conjugation (archaic)"]
    b"v2w-s" => VerbNidanWS,
    #[doc="Nidan verb (upper class) with 'yu' ending (archaic)"]
    b"v2y-k" => VerbNidanYK,
    #[doc="Nidan verb (lower class) with 'yu' ending (archaic)"]
    b"v2y-s" => VerbNidanYS,
    #[doc="Nidan verb (lower class) with 'zu' ending (archaic)"]
    b"v2z-s" => VerbNidanZS,
    #[doc="Yodan verb with 'bu' ending (archaic)"]
    b"v4b" => VerbYodanB,
    #[doc="Yodan verb with 'gu' ending (archaic)"]
    b"v4g" => VerbYodanG,
    #[doc="Yodan verb with 'hu/fu' ending (archaic)"]
    b"v4h" => VerbYodanH,
    #[doc="Yodan verb with 'ku' ending (archaic)"]
    b"v4k" => VerbYodanK,
    #[doc="Yodan verb with 'mu' ending (archaic)"]
    b"v4m" => VerbYodanM,
    #[doc="Yodan verb with 'nu' ending (archaic)"]
    b"v4n" => VerbYodanN,
    #[doc="Yodan verb with 'ru' ending (archaic)"]
    b"v4r" => VerbYodanR,
    #[doc="Yodan verb with 'su' ending (archaic)"]
    b"v4s" => VerbYodanS,
    #[doc="Yodan verb with 'tsu' ending (archaic)"]
    b"v4t" => VerbYodanT,
    #[doc="Godan verb - -aru special class"]
    b"v5aru" => VerbGodanAru,
    #[doc="Godan verb with 'bu' ending"]
    b"v5b" => VerbGodanB,
    #[doc="Godan verb with 'gu' ending"]
    b"v5g" => VerbGodanG,
    #[doc="Godan verb with 'ku' ending"]
    b"v5k" => VerbGodanK,
    #[doc="Godan verb - Iku/Yuku special class"]
    b"v5k-s" => VerbGodanKS,
    #[doc="Godan verb with 'mu' ending"]
    b"v5m" => VerbGodanM,
    #[doc="Godan verb with 'nu' ending"]
    b"v5n" => VerbGodanN,
    #[doc="Godan verb with 'ru' ending"]
    b"v5r" => VerbGodanR,
    #[doc="Godan verb with 'ru' ending (irregular verb)"]
    b"v5r-i" => VerbGodanRI,
    #[doc="Godan verb with 'su' ending"]
    b"v5s" => VerbGodanS,
    #[doc="Godan verb with 'tsu' ending"]
    b"v5t" => VerbGodanT,
    #[doc="Godan verb with 'u' ending"]
    b"v5u" => VerbGodanU,
    #[doc="Godan verb with 'u' ending (special class)"]
    b"v5u-s" => VerbGodanUS,
    #[doc="Godan verb - Uru old class verb (old form of Eru)"]
    b"v5uru" => VerbGodanUru,
    #[doc="intransitive verb"]
    b"vi" => VerbIntransitive,
    #[doc="Kuru verb - special class"]
    b"vk" => VerbKuru,
    #[doc="irregular nu verb"]
    b"vn" => VerbNu,
    #[doc="irregular ru verb, plain form ends with -ri"]
    b"vr" => VerbRu,
    #[doc="noun or participle which takes the aux. verb suru"]
    b"vs" => VerbSuru,
    #[doc="su verb - precursor to the modern suru"]
    b"vs-c" => VerbSu,
    #[doc="suru verb - included"]
    b"vs-i" => VerbSuruIncluded,
    #[doc="suru verb - special class"]
    b"vs-s" => VerbSuruSpecial,
    #[doc="transitive verb"]
    b"vt" => VerbTransitive,
    #[doc="Ichidan verb - zuru verb (alternative form of -jiru verbs)"]
    b"vz" => VerbIchidanZ,
);

jm_entity_enum!(
    JMReadingInfo;
    b"gikun" => Gikun,
    b"ik" => Irregular,
    b"ok" => Outdated,
    b"rk" => Rare,
    b"sk" => SearchOnly,
);
