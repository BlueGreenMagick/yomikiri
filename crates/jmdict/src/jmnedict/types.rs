use serde::{Deserialize, Serialize};

use crate::jmdict::JMSenseMisc;
use crate::utils::jm_entity_enum;

#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneDict {
    pub entries: Vec<JMneEntry>,
    pub creation_date: Option<String>,
}

/// `<entry>`
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneEntry {
    pub id: u32,
    /// 0+ k_ele
    pub kanjis: Vec<JMneKanji>,
    /// 1+ r_ele
    pub readings: Vec<JMneReading>,
    /// 1+ trans
    pub translations: Vec<JMneTranslation>,
}

/// `<k_ele>`
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneKanji {
    pub kanji: String,
    // Defined, but not used.
    // pub info: Vec<String>,
    pub priority: Vec<String>,
}

/// `<r_ele>`
#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMneReading {
    pub reading: String,
    pub to_kanji: Vec<String>,
    // Defined, but not used.
    // pub info: Vec<String>,
    pub priority: Vec<String>,
}

/// `<trans>`
///
/// At the date of development,
/// while specs allows translations in languages other than English,
/// no such translations actually exists in xml.
/// If non-English translations are added in the future,
/// they will be ignored and not be added to `translations` field.
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneTranslation {
    pub name_type: Vec<JMneNameType>,
    pub xref: Vec<String>,
    /// English translations are included
    pub translations: Vec<String>,
}

jm_entity_enum!(
    JMneNameType;
    #[doc="character"]
    b"char" => Character,
    #[doc="company name"]
    b"company" => Company,
    #[doc="creature"]
    b"creat" => Creature,
    #[doc="deity"]
    b"dei" => Deity,
    #[doc="document"]
    b"doc" => Document,
    #[doc="event"]
    b"ev" => Event,
    #[doc="female given name or forename"]
    #[serde(rename="forename (female)")]
    b"fem" => Female,
    #[doc="fiction"]
    b"fict" => Fiction,
    #[doc="given name or forename, gender not specified"]
    b"given" => Forename,
    #[doc="group"]
    b"group" => Group,
    #[doc="legend"]
    b"leg" => Legend,
    #[doc="male given name or forename"]
    #[serde(rename="forename (male)")]
    b"masc" => Male,
    #[doc="mythology"]
    b"myth" => Mythology,
    #[doc="object"]
    b"obj" => Object,
    #[doc="organization name"]
    b"organization" => Organization,
    #[doc="other"]
    b"oth" => Other,
    #[doc="full name of a particular person"]
    b"person" => Person,
    #[doc="place name"]
    b"place" => Place,
    #[doc="product name"]
    b"product" => Product,
    #[doc="religion"]
    b"relig" => Religion,
    #[doc="service"]
    b"serv" => Service,
    #[doc="ship name"]
    b"ship" => Ship,
    #[doc="railway station"]
    b"station" => Railway,
    #[doc="family or surname"]
    b"surname" => Surname,
    #[doc="unclassified name"]
    b"unclass" => Unclassified,
    #[doc="work of art, literature, music, etc. name"]
    b"work" => Artwork,
);

impl JMneReading {
    /// Returns true if reading is applicable to kanji
    ///
    /// `kanji` must be in same entry.
    pub fn is_for_kanji(&self, kanji: &String) -> bool {
        self.to_kanji.is_empty() || self.to_kanji.contains(kanji)
    }
}

impl From<JMneNameType> for JMSenseMisc {
    fn from(value: JMneNameType) -> Self {
        match value {
            JMneNameType::Artwork => JMSenseMisc::Artwork,
            JMneNameType::Character => JMSenseMisc::Character,
            JMneNameType::Company => JMSenseMisc::Company,
            JMneNameType::Creature => JMSenseMisc::Creature,
            JMneNameType::Deity => JMSenseMisc::Deity,
            JMneNameType::Document => JMSenseMisc::Document,
            JMneNameType::Event => JMSenseMisc::Event,
            JMneNameType::Female => JMSenseMisc::Female,
            JMneNameType::Fiction => JMSenseMisc::Fiction,
            JMneNameType::Forename => JMSenseMisc::Forename,
            JMneNameType::Group => JMSenseMisc::Group,
            JMneNameType::Legend => JMSenseMisc::Legend,
            JMneNameType::Male => JMSenseMisc::Male,
            JMneNameType::Mythology => JMSenseMisc::Mythology,
            JMneNameType::Object => JMSenseMisc::Object,
            JMneNameType::Organization => JMSenseMisc::Organization,
            JMneNameType::Other => JMSenseMisc::Other,
            JMneNameType::Person => JMSenseMisc::Person,
            JMneNameType::Place => JMSenseMisc::Place,
            JMneNameType::Product => JMSenseMisc::Product,
            JMneNameType::Railway => JMSenseMisc::Railway,
            JMneNameType::Religion => JMSenseMisc::Religion,
            JMneNameType::Service => JMSenseMisc::Service,
            JMneNameType::Ship => JMSenseMisc::Ship,
            JMneNameType::Surname => JMSenseMisc::Surname,
            JMneNameType::Unclassified => JMSenseMisc::Unclassified,
        }
    }
}
