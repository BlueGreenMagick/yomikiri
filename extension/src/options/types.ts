export const options: OptionsGroup[] = [
  {
    name: "Anki",
    options: [
      {
        type: "number",
        key: "anki.connect_port",
        title: "AnkiConnect port number",
        description: "This is the AnkiConnect config `webBindPort`.",
      },
      {
        type: "click",
        title: "Setup Anki template",
        onClick: () => {
          console.log("Setup Anki Template");
        },
      },
    ],
  },
];

export interface OptionsGroup {
  name: string;
  options: Option[];
}

export type Option = ClickOptions | NumberOptions;

interface BaseOption {
  title: string;
  description?: string;
}

export interface ClickOptions extends BaseOption {
  type: "click";
  onClick: () => void;
}

export interface NumberOptions extends BaseOption {
  key: string;
  type: "number";
}
