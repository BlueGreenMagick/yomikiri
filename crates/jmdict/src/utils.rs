/*
To easily create jm entity enums,
run below code on xml entity declarations.

```js
function macroCode(decl) {
    let output = ""
    for (const m of decls.matchAll(/<!ENTITY ([^\s]+) "([^"]*)">/g)) {
        output += `#[doc="${m[2]}"]`
        output += 'b"'
        output += m[1]
        output += '" => '
        const firstWord =  m[2].split(" ")[0]
        output += firstWord[0].toUpperCase() + firstWord.slice(1)
        output += ",\n"
    }
    return output
}

let decls = `<Paste entity declarations in xml>`
console.log(macroCode(decls))
```
*/

macro_rules! jm_entity_enum {
  (
      $( #[$enum_attrs:meta] )*
      $enum_name:ident;
      $(
        $( #[$attrs:meta] )*
        $( $key:pat $(if $guard:expr)? )? => $variant:ident
      ),+,
  ) => {
      #[cfg_attr(feature = "wasm", derive(::tsify_next::Tsify))]
      #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
      #[serde(rename_all = "camelCase")]
      $( #[$enum_attrs] )*
      pub enum $enum_name {
          $(
              $( #[$attrs] )*
              $variant,
          )+
      }

      impl $enum_name {
          pub fn parse_field(field: &[u8]) -> Option<$enum_name> {
              if is_single_entity(field) {
                  match &field[1..field.len() -1] {
                      $(
                          $(
                            $key $(if $guard)? => Some(Self::$variant),
                          )?
                      )+
                      _ => None,
                  }
              } else {
                  None
              }

          }
      }
  };
}

pub(crate) use jm_entity_enum;
