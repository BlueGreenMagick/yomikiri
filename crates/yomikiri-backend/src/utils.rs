#[cfg(uniffi)]
use std::time::{SystemTime, UNIX_EPOCH};

#[cfg(wasm)]
pub fn set_panic_hook() {
    console_error_panic_hook::set_once();
}

#[cfg(wasm)]
pub fn setup_logger() {
    wasm_logger::init(wasm_logger::Config::default());
}

#[cfg(not(any(wasm, apple)))]
pub fn setup_logger() {
    eprintln!("No logger configurable for targets not wasm nor apple")
}

#[cfg(wasm)]
#[allow(dead_code)]
pub(crate) fn time_now() -> f64 {
    let window = web_sys::window().unwrap();
    window.performance().unwrap().now()
}

#[cfg(uniffi)]
#[allow(dead_code)]
pub(crate) fn time_now() -> f64 {
    let micro = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_micros();
    (micro as f64) / 1000.0
}
