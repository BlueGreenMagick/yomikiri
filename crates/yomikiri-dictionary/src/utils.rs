use std::borrow::Cow;

use unicode_normalization::{is_nfkc, UnicodeNormalization};

pub fn nfkc_normalize<'a, S: Into<Cow<'a, str>>>(text: S) -> Cow<'a, str> {
    let text: Cow<'_, str> = text.into();

    if is_nfkc(&text) {
        text
    } else {
        Cow::Owned(text.nfkc().collect())
    }
}
