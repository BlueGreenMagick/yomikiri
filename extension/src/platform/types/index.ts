export interface Module {
  IS_DESKTOP: boolean;
  IS_IOS: boolean;
}
