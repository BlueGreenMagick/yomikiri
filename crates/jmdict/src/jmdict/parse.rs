use core::str;
use std::io::BufRead;

use log::warn;
use quick_xml::events::Event;
use quick_xml::Reader;

use super::types::{
    JMDialect, JMDict, JMEntry, JMKanji, JMKanjiInfo, JMPartOfSpeech, JMReading, JMSense,
    JMSenseMisc,
};
use crate::jmdict::types::JMReadingInfo;
use crate::utils::parse_entity_enum_into;
use crate::xml::{get_next_child_in, parse_string_in_tag_into, TagName};
use crate::{Error, Result};

struct JMDictParser<R: BufRead> {
    reader: Reader<R>,
    buf: Vec<u8>,
}

impl<R: BufRead> JMDictParser<R> {
    pub fn new(reader: R) -> Result<Self> {
        let mut reader = Reader::from_reader(reader);
        reader.config_mut().trim_text_start = true;
        reader.config_mut().trim_text_end = true;
        reader.config_mut().expand_empty_elements = true;

        let buf = Vec::new();
        let mut parser = JMDictParser { reader, buf };
        parser.parse_jmdict_start()?;
        Ok(parser)
    }

    fn parse_jmdict_start(&mut self) -> Result<()> {
        loop {
            self.buf.clear();
            match self.reader.read_event_into(&mut self.buf)? {
                Event::Start(tag) => match tag.name().0 {
                    b"JMdict" => return Ok(()),
                    _ => {
                        warn!("Unknown global tag: {}", tag.tag_name());
                    }
                },
                Event::Eof => return Err(Error::InvalidXml("<JMdict> not found".into())),
                _ => {}
            }
        }
    }

    fn parse_entry_in_jmdict(&mut self) -> Result<Option<JMEntry>> {
        loop {
            let start = match get_next_child_in(&mut self.reader, &mut self.buf, "JMdict")? {
                Some(tag) => tag,
                None => return Ok(None),
            };
            match start.name().0 {
                b"entry" => return self.parse_in_entry().map(Some),
                _ => {
                    warn!("Unknown tag in <jmdict>: <{}>", start.tag_name());
                }
            }
        }
    }

    fn parse_in_entry(&mut self) -> Result<JMEntry> {
        let mut id: Option<u32> = None;
        let mut entry = JMEntry::default();

        while let Some(start) = get_next_child_in(&mut self.reader, &mut self.buf, "entry")? {
            match start.name().0 {
                b"ent_seq" => {
                    if let Some(id) = id {
                        return Err(Error::MultipleEntryIds(id));
                    }
                    let idstr =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ent_seq")?;
                    let seq_id = str::parse::<u32>(&idstr)
                        .map_err(|_| format!("Couldn't parse as u32 number: {}", idstr))?;
                    id = Some(seq_id);
                }
                b"k_ele" => {
                    let form = self.parse_in_form()?;
                    entry.kanjis.push(form);
                }
                b"r_ele" => {
                    let reading = self.parse_in_reading()?;
                    entry.readings.push(reading);
                }
                b"sense" => {
                    let sense = self.parse_in_sense()?;
                    entry.senses.push(sense);
                }
                _ => {
                    warn!("Unknown tag in <entry>: <{}>", start.tag_name())
                }
            };
        }

        if let Some(id) = id {
            entry.id = id;
            return Ok(entry);
        } else {
            return Err(Error::NoEntryId(self.reader.buffer_position()));
        }
    }

    fn parse_in_form(&mut self) -> Result<JMKanji> {
        let mut form = JMKanji::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "k_ele")? {
            match tag.name().0 {
                b"keb" => {
                    if !form.kanji.is_empty() {
                        warn!("Found multiple <keb> in form '{}' in JMDict", form.kanji)
                    }
                    form.kanji = parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"keb")?;
                }
                b"ke_inf" => {
                    parse_entity_enum_into!(
                        JMKanjiInfo,
                        &mut self.reader,
                        &mut self.buf,
                        "ke_inf",
                        form.info
                    );
                }
                b"ke_pri" => {
                    form.priority.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"ke_pri",
                    )?);
                }
                _ => {
                    warn!("Unknown tag in <entry>: {}", tag.tag_name());
                }
            };
        }

        Ok(form)
    }

    fn parse_in_reading(&mut self) -> Result<JMReading> {
        let mut reading = JMReading::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "r_ele")? {
            match tag.name().0 {
                b"reb" => {
                    reading.reading =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"reb")?;
                }
                b"re_nokanji" => {
                    reading.nokanji = true;
                    // consume ending tag
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"re_nokanji")?;
                }
                b"re_restr" => {
                    reading.to_form.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"re_restr",
                    )?);
                }
                b"re_inf" => {
                    parse_entity_enum_into!(
                        JMReadingInfo,
                        &mut self.reader,
                        &mut self.buf,
                        "re_inf",
                        reading.info
                    );
                }
                b"re_pri" => {
                    reading.priority.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"re_pri",
                    )?);
                }
                _ => {
                    warn!("Unknown tag in <r_ele>: {}", &tag.tag_name());
                }
            };
        }

        Ok(reading)
    }

    fn parse_in_sense(&mut self) -> Result<JMSense> {
        let mut sense = JMSense::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "sense")? {
            match tag.name().0 {
                b"stagk" => {
                    sense.to_form.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"stagk",
                    )?);
                }
                b"stagr" => {
                    sense.to_reading.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"stagr",
                    )?);
                }
                b"pos" => {
                    parse_entity_enum_into!(
                        JMPartOfSpeech,
                        &mut self.reader,
                        &mut self.buf,
                        "pos",
                        sense.pos
                    );
                }
                b"xref" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"xref")?;
                }
                b"ant" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ant")?;
                }
                b"field" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"field")?;
                }
                b"misc" => {
                    parse_entity_enum_into!(
                        JMSenseMisc,
                        &mut self.reader,
                        &mut self.buf,
                        "misc",
                        sense.misc
                    );
                }
                b"s_inf" => {
                    sense.info.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"s_inf",
                    )?);
                }
                b"dial" => {
                    parse_entity_enum_into!(
                        JMDialect,
                        &mut self.reader,
                        &mut self.buf,
                        "dial",
                        sense.dialects
                    );
                }
                b"gloss" => {
                    sense.meanings.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"gloss",
                    )?);
                }
                b"lsource" => {
                    // consume ending tag
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"lsource")?;
                }
                b"example" => {
                    parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"example")?;
                }
                _ => {
                    warn!("Unknown tag in <sense>: {}", tag.tag_name());
                }
            };
        }

        Ok(sense)
    }
}

pub fn parse_jmdict_xml(xml: &str) -> Result<JMDict> {
    let mut parser = JMDictParser::new(xml.as_bytes())?;
    let mut entries = vec![];
    while let Some(entry) = parser.parse_entry_in_jmdict()? {
        entries.push(entry);
    }
    Ok(JMDict { entries })
}
