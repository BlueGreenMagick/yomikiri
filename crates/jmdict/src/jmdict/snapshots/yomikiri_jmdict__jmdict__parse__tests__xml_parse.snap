---
source: crates/jmdict/src/jmdict/parse.rs
assertion_line: 260
expression: result
---
entries:
  - id: 1000040
    forms:
      - form: 〃
        info: []
        priority: []
    readings:
      - reading: おなじ
        nokanji: false
        to_form: []
        info: []
        priority: []
      - reading: おなじく
        nokanji: false
        to_form: []
        info: []
        priority: []
    senses:
      - to_form: []
        to_reading: []
        part_of_speech:
          - "=n="
        misc: []
        info: []
        dialect: []
        meaning:
          - ditto mark
  - id: 1000050
    forms:
      - form: 仝
        info: []
        priority: []
    readings:
      - reading: どうじょう
        nokanji: false
        to_form: []
        info: []
        priority: []
    senses:
      - to_form: []
        to_reading: []
        part_of_speech:
          - "=n="
        misc: []
        info: []
        dialect: []
        meaning:
          - "\"as above\" mark"
